topdown/reachable.go

```diff
@@ -56,6 +69,67 @@ func builtinReachable(bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term
 	return iter(ast.NewTerm(reached))
 }
 
+// pathBuilder is called recursively to build an array of paths that are reachable from the root
+func pathBuilder(graph ast.Object, root *ast.Term, path []*ast.Term, paths []*ast.Term, reached ast.Set) []*ast.Term {
+	if edges := graph.Get(root); edges != nil {
+		path = append(path, root)
+
+		if numberOfEdges(edges) >= 1 {
+			foreachVertex(edges, func(neighbor *ast.Term) {
+				if reached.Contains(neighbor) {
+					// If we've already reached this node, return current path (avoid infinite recursion)
+					paths = append(paths, path...)
+				} else {
+					reached.Add(root)
+					paths = pathBuilder(graph, neighbor, path, paths, reached)
+				}
+			})
+		} else {
+			paths = append(paths, path...)
+		}
+	}
+
+	return paths
+}
+
+func builtinReachablePaths(bctx BuiltinContext, args []*ast.Term, iter func(*ast.Term) error) error {
+	// Return the empty set if the first argument is not an object.
+	graph, err := builtins.ObjectOperand(args[0].Value, 1)
+	if err != nil {
+		return err
+	}
+
+	// This is a queue that holds all nodes we still need to visit.  It is
+	// initialised to the initial set of nodes we start out with.
+	var queue []*ast.Term
+	foreachVertex(args[1], func(t *ast.Term) {
+		queue = append(queue, t)
+	})
+
+	results := ast.NewSet()
+
+	for len(queue) > 0 {
+		node := queue[0]
+
+		// Find reachable paths from edges in root node in queue and append arrays to the results set
+		if edges := graph.Get(node); edges != nil {
```

Comment:
This looks remarkably similar to `pathBuilder`... but we can't use that here "one level up", can we? (I'm thinking about avoiding the duplication...

Reply:
Yeah, it is remarkably similar, but we can't use it one level up :(